query-service-impl/src/main/java/org/hypertrace/core/query/service/attribubteexpression/AttributeExpressionSubpathExistsFilteringTransformation.java

```diff
@@ -0,0 +1,134 @@
+
+    Set<Filter> rootFilters =
+        ImmutableSet.<Filter>builder()
+            .add(updateFilterForComplexAttributeExpressionFromFilter(originalFilter))
+            .addAll(createFilterForComplexAttributeExpressionFromOrderBy(orderBys))
+            .addAll(createFilterForComplexAttributeExpressionFromSelection(selections))
+            .build()
+            .stream()
+            .filter(Predicate.not(Filter.getDefaultInstance()::equals))
+            .collect(ImmutableSet.toImmutableSet());
+
+    if (rootFilters.isEmpty()) {
+      return Optional.empty();
+    }
+
+    if (rootFilters.size() == 1) {
+      return rootFilters.stream().findFirst();
+    }
+    return Optional.of(
+        Filter.newBuilder().setOperator(Operator.AND).addAllChildFilter(rootFilters).build());
+  }
+
+  private Filter updateFilterForComplexAttributeExpressionFromFilter(Filter originalFilter) {
+    /*
+     * If childFilter is present, then the expected operators comprise the logical operators.
+     * If childFilter is absent, then the filter is a leaf filter which will have lhs and rhs.
+     */
+    if (originalFilter.getChildFilterCount() > 0) {
+      Filter.Builder builder = Filter.newBuilder();
+      builder.setOperator(originalFilter.getOperator());
+      originalFilter
+          .getChildFilterList()
+          .forEach(
+              childFilter ->
+                  builder.addChildFilter(
+                      updateFilterForComplexAttributeExpressionFromFilter(childFilter)));
+      return builder.build();
+    }
+    if (isAttributeExpressionWithSubpath(originalFilter.getLhs())) {
+      Filter childFilter =
+          createContainsKeyFilter(originalFilter.getLhs().getAttributeExpression());
+      return Filter.newBuilder()
+          .setOperator(Operator.AND)
+          .addChildFilter(originalFilter)
+          .addChildFilter(childFilter)
+          .build();
+    }
+    return originalFilter;
+  }
+
+  private List<Filter> createFilterForComplexAttributeExpressionFromOrderBy(
+      List<OrderByExpression> orderByExpressionList) {
+    return orderByExpressionList.stream()
+        .map(OrderByExpression::getExpression)
+        .filter(QueryRequestUtil::isAttributeExpressionWithSubpath)
+        .map(Expression::getAttributeExpression)
+        .map(QueryRequestUtil::createContainsKeyFilter)
+        .collect(Collectors.toList());
+  }
+
+  private List<Filter> createFilterForComplexAttributeExpressionFromSelection(
+      List<Expression> expressions) {
+    return expressions.stream()
+        .flatMap(this::getAnyAttributeExpression)
+        .map(QueryRequestUtil::createContainsKeyFilter)
+        .collect(Collectors.toList());
+  }
+
+  private Stream<AttributeExpression> getAnyAttributeExpression(Expression expression) {
+    if (expression.hasFunction()) {
+      return expression.getFunction().getArgumentsList().stream()
+          .flatMap(this::getAnyAttributeExpression);
+    }
+    return Stream.of(expression)
```

Comment:
nit : Can we have this in else block ?

Reply:
style preference - I always unwrap else blocks if the if statement returns to reduce nesting. Here's a lint rule (for a different language, but same pattern) that gives a tad more context: https://palantir.github.io/tslint/rules/unnecessary-else/